### ISSUE TITLE
[rt.openssl.org #4681] X.509 load method

### ISSUE THREAD

**Issue by richsalz** [openssl/openssl#2531]
Migrated from [rt.openssl.org#4681](https://rt.openssl.org/Ticket/Display.html?id=4681) (status was 'new')

Requestors:
* openssl@roumenpetrov.info

From openssl@roumenpetrov.info on 2016-09-22 12:22:59:
```
This is an enhancement request.

OpenSSL 1.1 hides details of structures used to load X.509 certificates, 
in particular - x509_lookup_method_st , x509_lookup_st and x509_object_st.
This impact non OpenSSL projects as external application has to 
duplicated those structures.

Request is OpenSSL do not change those structures until new 
implementation in a binary incompatible release.


It seems to me current look-up method is quite complex.
For instance get_by_subject performs two main steps load and query - see 
"by_dir".
In first step code using "query constraint" fetch data and adds each 
item found to store. Update of store requires thread lock.
In second step code query entire store using "query constraint" and 
prepare result. Query of store requires thread lock.

I guess that could be used more simple "callback" model - a method with 
callback context.
Method fetch data, return only item (certificate, CRL, etc.) on each 
call and so until end of data.
With this model library is responsible to update store and to prepare 
result. Lock of store could be managed internally.


Regards,
Roumen

```

**Comment by kaduk** [comment]
I don't understand the request ... 1.1 is a binary-incompatible release, and the structures are intentionally opaque so that they can be changed if needed.
The second half of the post seems on a different topic, and is probably too speculative to keep open without a concrete patch/proposal.
So, I would say this should be closed.

**Comment by levitte** [comment]
I agree.

**Comment by petrovr** [comment]
Let me rephrase -How to implement X5.09 lookup method with 1.1+ ?

**Comment by kaduk** [comment]
At present, the answer is "patch the source".
And to some extent it ought to be, as the lookup ->new() method has to modify the X509_LOOKUP object that is also an opaque type.  (Yes, there are controls funtions and such, but that's pretty grotty.)
So, with that clarification, there is an actual issue here, yes.
Out of curiousity, what sort of lookup methods do you want to make?

**Comment by petrovr** [comment]
kaduk wrote:
> At present, the answer is "patch the source".
I did not understand you.

> Out of curiousity, what sort of lookup methods do you want to make?
Actually to port existing by_ldap to new API.
I was written long time ago:
*[24 Nov 2004 :* Version 
x509-5.1](http://roumenpetrov.info/openssh/index-20070310.html)
* X.509 store" now can use LDAP to lookup for certificates and CRLs.

In PKIX-SSH are used following methods by_file, by_dir and by_ldap.

PKIX-SSH is modified to use OpenSSL API 1.1 and has backward compatible 
function definitions.

For X.509 lookup, I have no choice except to define locally in 
x509_by_ldap.c structures hidden in 1.1.

Roumen

**Comment by richsalz** [comment]
@kaduk meant that it is not possible without modifying openssl source.

But we feel this is a bug caused by making things opaque and we should fix it.  Let's talk on openssl-dev about the minimal set of changes needed to be able to write your own lookup routines.

**Comment by petrovr** [comment]
Rich Salz wrote:
> @kaduk meant that it is not possible without modifying openssl source.
Yes
> But we feel this is a bug caused by making things opaque and we should fix it.
Yes
> Let's talk on openssl-dev about the minimal set of changes needed to be able to write your own lookup routines.
Minimal depend from point of view.
If there is "store"-functionality it could be used as replacement of  
lookup-method as I already point in openssh-dev list.

Roumen

**Comment by richsalz** [comment]
So you think the upcoming store stuff (being done by @levitte ) will be enough, and close this?

**Comment by levitte** [comment]
I've been looking at this stuff, and am pondering a X509_LOOKUP_METHOD that's wrapped around the STORE stuff.  I'll try to have a closer look at that next week.

**Comment by levitte** [comment]
For anyone interested, have a look at [the STORE project](https://github.com/openssl/openssl/projects/1).  Quite a bit has happened, and for the moment, I'm slowly looking at the X509_LOOKUP and X509_STORE stuff to see what can be done further.

**Comment by mattcaswell** [comment]
Ping @levitte - can this be closed now?

**Comment by mattcaswell** [comment]
Ping @levitte.
